app/services/puppet_fact_parser.rb

```diff
@@ -144,8 +144,25 @@ def disks_total
     facts['disks']&.values&.sum { |disk| disk&.fetch('size_bytes', 0).to_i }
   end
 
+  def foreman_uuid
+    signed_fact('foreman_uuid')
+  end
+
+  def smart_proxy_uuid
+    signed_fact('smart_proxy_uuid')
+  end
+
   private
 
+  def signed_fact(key)
+    uuid = facts[key]
+    signature = facts["#{key}_signature"]
+
+    if uuid.present? && signature.present? && signature == OpenSSL::HMAC.hexdigest("SHA512", SETTINGS[:oauth_consumer_secret], uuid)
+      uuid
+    end
+  end
```

Comment:
Nitpick: this will return `false` on incorrect signature. Not sure if this was intention.

Reply:
Will it? The intention was to return `nil`, which as far as I can tell it does. Or am I missing something?

Reply:
You are right, okay.

```
?> if false
>> end
=> nil
```